### PR TITLE
Implement Downstream TcpKeepalive

### PR DESCRIPTION
**Pull request by arkodg**
Relates to https://github.com/envoyproxy/gateway/issues/1031

### CODE REVIEW

File: internal/ir/xds.go [hunk 1]
```diff
@@ -934,5 +938,18 @@ type Tracing struct {
...
-	EnableVirtualHostStats bool `json:"enableVirtualHostStats"`
+	EnableVirtualHostStats bool `json:"enableVirtualHostStats" yaml:"enableVirtualHostStats"`
+}
+
+// TCPKeepalive define the TCP Keepalive configuration.
+// +k8s:deepcopy-gen=true
+type TCPKeepalive struct {
+	// The total number of unacknowledged probes to send before deciding
+	// the connection is dead.
+	Probes *uint32 `json:"probes,omitempty" yaml:"probes,omitempty"`
```
Comment (qicz) [inline internal/ir/xds.go:949]: suggestion: provide initial values as a reference

File: internal/xds/translator/tcpkeepalive.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+func buildTCPKeepaliveSocketOptions(keepAlive *ir.TCPKeepalive) []*corev3.SocketOption {
+	if keepAlive == nil {
+		return nil
+	}
+
+	socketOptions := make([]*corev3.SocketOption, 0)
+	// Enable Keep Alives
+	socketOption := &corev3.SocketOption{
+		Description: "socket option to enable tcp keep alive",
+		Level:       0x1,                                        // syscall.SOL_SOCKET has a different value for Darwin, resulting in `go test` failing
```
Comment (zirain) [inline internal/xds/translator/tcpkeepalive.go:24]: do you means `go test` will failing when someone coding with mac?

Comment (arkodg) [reply internal/xds/translator/tcpkeepalive.go:24]: yes, this is from my macbook
```
git diff
diff --git a/internal/xds/translator/tcpkeepalive.go b/internal/xds/translator/tcpkeepalive.go
index ebf129ca..07201803 100644
--- a/internal/xds/translator/tcpkeepalive.go
+++ b/internal/xds/translator/tcpkeepalive.go
@@ -6,6 +6,8 @@
 package translator
 
 import (
+       "syscall"
+
        corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
 
        "github.com/envoyproxy/gateway/internal/ir"
@@ -21,8 +23,8 @@ func buildTCPKeepaliveSocketOptions(keepAlive *ir.TCPKeepalive) []*corev3.Socket
        // Enable Keep Alives
        socketOption := &corev3.SocketOption{
                Description: "socket option to enable tcp keep alive",
-               Level:       0x1,                                        // syscall.SOL_SOCKET has a different value for Darwin, resulting in `go test` failing
-               Name:        0x9,                                        // syscall.SO_KEEPALIVE has a different value for Darwin, resulting in `go test` failing
+               Level:       syscall.SOL_SOCKET,                         // syscall.SOL_SOCKET has a different value for Darwin, resulting in `go test` failing
+               Name:        syscall.SO_KEEPALIVE,                       // syscall.SO_KEEPALIVE has a different value for Darwin, resulting in `go test` failing
                Value:       &corev3.SocketOption_IntValue{IntValue: 1}, // Enable
        }
 
🐳 ~/go-workspace/src/github.com/envoyproxy/gateway/internal/xds/translator$ make test^C
🐳 ~/go-workspace/src/github.com/envoyproxy/gateway/internal/xds/translator$ go test ./...
--- FAIL: TestTranslateXds (0.05s)
    --- FAIL: TestTranslateXds/listener-tcp-keepalive (0.00s)
        translator_test.go:218: 
            	Error Trace:	/Users/arkodebdasgupta/go-workspace/src/github.com/envoyproxy/gateway/internal/xds/translator/translator_test.go:218
            	Error:      	Not equal: 

            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -36,4 +36,4 @@
            	            	     intValue: "1"
            	            	-    level: "1"
            	            	-    name: "9"
            	            	+    level: "65535"
            	            	+    name: "8"
            	            	 - address:
            	            	@@ -74,4 +74,4 @@
            	            	     intValue: "1"
            	            	-    level: "1"
            	            	-    name: "9"
            	            	+    level: "65535"
            	            	+    name: "8"
            	            	   - description: socket option for keep alive probes
            	            	@@ -111,4 +111,4 @@
            	            	     intValue: "1"
            	            	-    level: "1"
            	            	-    name: "9"
            	            	+    level: "65535"
            	            	+    name: "8"
            	            	 - address:
            	            	@@ -129,4 +129,4 @@
            	            	     intValue: "1"
            	            	-    level: "1"
            	            	-    name: "9"
            	            	+    level: "65535"
            	            	+    name: "8"
            	            	   - description: socket option for keep alive probes
            	Test:       	TestTranslateXds/listener-tcp-keepalive
FAIL
FAIL	github.com/envoyproxy/gateway/internal/xds/translator	0.562s
ok  	github.com/envoyproxy/gateway/internal/xds/translator/runner	1.118s
FAIL
```

File: internal/ir/xds.go (hunk 1 above)
Comment (arkodg) [reply internal/ir/xds.go:950]: done

File: internal/xds/translator/tcpkeepalive.go (hunk 2 above)
Comment (zirain) [reply internal/xds/translator/tcpkeepalive.go:24]: prefer to make this const, and have detail comment on the top on this file.

Comment (arkodg) [reply internal/xds/translator/tcpkeepalive.go:24]: But the const won't work, as shared in above test

File: internal/xds/translator/translator.go [hunk 3]
```diff
@@ -291,7 +291,7 @@ func processTCPListenerXdsTranslation(tCtx *types.ResourceVersionTable, tcpListe
 		// Search for an existing listener, if it does not exist, create one.
 		xdsListener := findXdsListenerByHostPort(tCtx, tcpListener.Address, tcpListener.Port, corev3.SocketAddress_TCP)
 		if xdsListener == nil {
-			xdsListener = buildXdsTCPListener(tcpListener.Name, tcpListener.Address, tcpListener.Port, accesslog)
+			xdsListener = buildXdsTCPListener(tcpListener.Name, tcpListener.Address, tcpListener.Port, tcpListener.TCPKeepalive, accesslog)
```
Comment (zirain) [inline internal/xds/translator/translator.go:294]: can you add a TODO here, this function has too much parameters, four of them come from `tcpListener`.

File: internal/xds/translator/tcpkeepalive.go (hunk 2 above)
Comment (zirain) [reply internal/xds/translator/tcpkeepalive.go:24]: I will take a look at later, this should not be a blocker.

File: internal/xds/translator/tcpkeepalive.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+package translator
+
+import (
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+
+	"github.com/envoyproxy/gateway/internal/ir"
+)
+
+// buildTCPKeepaliveSocketOptions converts listener downstream tcpKeepAlive settings to xds socketOptions
+func buildTCPKeepaliveSocketOptions(keepAlive *ir.TCPKeepalive) []*corev3.SocketOption {
```
Comment (zirain) [inline internal/xds/translator/tcpkeepalive.go:15]: rename this to `buildTCPSocketOptions`, `keepalive` is part of `SocketOption`

**Review by zirain — APPROVED** [review]
LGTM, small nits.

**Review by zhaohuabing — APPROVED** [review]
LGTM

**Review by zirain — APPROVED** [review]

**Pull request merged**